### ISSUE TITLE
Variational Compress Method

### ISSUE THREAD

**Issue by jjren** [shuaigroup/Renormalizer#83]
Rewrite the variational compress method in the old ephMPS in Renormalizer.
MPO \times MPS with a large bond dimension consumes a lot of memory. Variational compress method could help.

**Comment by liwt31** [comment]
I'm looking forward to this method because in charge transport calculation J @ ket is quite time & memory consuming.
The variational method should help with both accuracy and efficiency.

And a suggestion about the initial guess: SVD compress MPO \times MPS without QR canonicalisation. The compression would be tricky to implement because we would like to avoid explicitly construct the whole MPO \times MPS.

**Comment by jjren** [comment]
I will commit it this week.

**Comment by jjren** [comment]
> And a suggestion about the initial guess: SVD compress MPO \times MPS without QR canonicalisation. The compression would be tricky to implement because we would like to avoid explicitly construct the whole MPO \times MPS.

I tried a similar algorithm (Ann. Phys. 2011, 326, 96−192.) in the old ephMPS code, in which the initial guess is  `canonicalised mpo @ canonicalised mps` and then compressed without full canonicalizsation.
I remembered it does not improve much compared to a random guess but make the code messy. 
Thus in #98 , the initial guess is constructed from compressed_mpo @ compressed_mps and each of them has `M=5`.
